TODO-rebase.sh

```diff
@@ -0,0 +1,738 @@
+#!/bin/sh
+#
+# Copyright (c) 2005 Junio C Hamano.
+#
+
+SUBDIRECTORY_OK=Yes
+OPTIONS_KEEPDASHDASH=
+OPTIONS_STUCKLONG=t
+OPTIONS_SPEC="\
+# git rebase [-i] [options] [--exec <cmd>] [--onto <newbase>] [<upstream>] [<branch>]
+# git rebase [-i] [options] [--exec <cmd>] [--onto <newbase>] --root [<branch>]
+# git rebase --continue | --abort | --skip | --edit-todo
+--
+ Available options are
+v,verbose!         display a diffstat of what changed upstream
+q,quiet!           be quiet. implies --no-stat
+autostash          automatically stash/stash pop before and after
+fork-point         use 'merge-base --fork-point' to refine upstream
+#onto=!             rebase onto given branch instead of upstream
+r,rebase-merges?   try to rebase merges instead of skipping them
+p,preserve-merges! try to recreate merges instead of ignoring them
+s,strategy=!       use the given merge strategy
+no-ff!             cherry-pick all commits, even if unchanged
+m,merge!           use merging strategies to rebase
+i,interactive!     let the user edit the list of commits to rebase
+x,exec=!           add exec lines after each commit of the editable list
+k,keep-empty	   preserve empty commits during rebase
+allow-empty-message allow rebasing commits with empty messages
+f,force-rebase!    force rebase even if branch is up to date
+X,strategy-option=! pass the argument through to the merge strategy
+stat!              display a diffstat of what changed upstream
+n,no-stat!         do not show diffstat of what changed upstream
+verify             allow pre-rebase hook to run
+rerere-autoupdate  allow rerere to update index with resolved conflicts
+root!              rebase all reachable commits up to the root(s)
+autosquash         move commits that begin with squash!/fixup! under -i
+committer-date-is-author-date! passed to 'git am'
+ignore-date!       passed to 'git am'
+signoff            passed to 'git am'
+whitespace=!       passed to 'git apply'
+ignore-whitespace! passed to 'git apply'
+C=!                passed to 'git apply'
+S,gpg-sign?        GPG-sign commits
+ Actions:
+# continue!          continue
+abort!             abort and check out the original branch
+skip!              skip current patch and continue
+edit-todo!         edit the todo list during an interactive rebase
+quit!              abort but keep HEAD where it is
+show-current-patch! show the patch file being applied or merged
+"
+. git-sh-setup
+set_reflog_action rebase
+require_work_tree_exists
+cd_to_toplevel
+
+LF='
+'
+ok_to_skip_pre_rebase=
+resolvemsg="
```

Comment:
After staring at `git-rebase--am.sh` for a while, I realized that this `resolvemsg` variable is never used in `git-rebase.sh` directly, but in the rebase backends, and therefore needs to be moved to `git-rebase--common.sh`.

Reply:
To make this a bit more complete, I ran a command-line to figure out which variables are used by the `am` backend:

```sh
$ sed 's/\$/\n&/g' <git-rebase--am.sh | sed -n 's/\(\$[a-zA-Z_][0-9a-zA-Z_]*\).*/\1/p' | sort | uniq
$GIT_DIR
$action
$allow_rerere_autoupdate
$git_am_opt
$git_format_patch_opt
$gpg_sign_opt
$head_name
$onto
$orig_head
$rebase_root
$resolvemsg
$restrict_revision
$ret
$revisions
$state_dir
$upstream
```

Out of these, `ret` is definitely defined in the backend itself. The others are defined via `git-rebase.sh` (`git-rebase--am.sh` re-defines, quite unnecessarily, `revisions`, but it does not matter, we have to keep defining it in `run_specific_rebase()` for the other backends).

Out of the remaining variables, we already define these in the `script_snippet`:

```
$GIT_DIR
$action
$git_am_opt
$head_name
$onto
$orig_head
$revisions
$state_dir
$upstream
```

That leaves these variables that we definitely still have to define in the end (read: once the conversion is complete):

```
$allow_rerere_autoupdate
$git_format_patch_opt
$gpg_sign_opt
$rebase_root
$restrict_revision
``

Note: the `restrict_revision` variable is `unset` in `git-rebase.sh`, so we will have to replicate this (probably via handling `NULL` being passed as `value` to `add_var()` to imply that the variable should be `unset` rather than defined).

Reply:
Running the same analysis for `git-rebase--merge.sh`, the following variables need to be handled, in addition:

```
$allow_empty_message
$signoff
$strategy
$strategy_opts
```

(The command-line prints quite a few more variables, but they are defined *within* `git-rebase--merge.sh`, so there is no need for us to define them in `run_specific_rebase()` via that `script_snippet`.)

Reply:
For `git-rebase--interactive.sh`, we also need these, in addition to the previously mentioned ones:

```
$GIT_REFLOG_ACTION
$autosquash
$cmd
$force_rebase
$keep_empty
$squash_onto
$switch_to
```

Reply:
Finally, `git-rebase--preserve-merges.sh` is quite a bit more tricky to analyze because it uses *so* many variables. So I used this command line to match up the variables used in `git-rebase--preserve-merges.sh` that are also used in assignments in `git-legacy-rebase.sh`:

```sh
(sed 's/\$/\n&/g' <git-rebase--preserve-merges.sh |
   sed -n 's/\(\$[a-zA-Z_][0-9a-zA-Z_]*\).*/\1/p' | sort | uniq;
 sed 's/=/&\n/g' <git-legacy-rebase.sh |
   sed -ne 's/.*[^a-zA-Z0-9_=]//' -e 's/\([a-zA-Z_][0-9a-zA-Z_]*\)=.*/$\1/p' | sort | uniq) |
sort | uniq -d
```

And the only remaining variable we need to pass (eventually) via the `script_snippet` is `verbose`. Phew. That's it.